kinto/core/resource/__init__.py

```diff
@@ -312,28 +331,47 @@ def plural_get(self):
 
         pagination_rules, offset = self._extract_pagination_rules_from_token(limit, sorting)
 
-        objects, total_objects = self.model.get_objects(
+        # The reason why we call self.model.get_objects() with `limit=limit + 1` is to avoid
+        # having to count the total number of objects in the database just to be able
+        # to *decide* whether or not to have a `Next-Page` header.
+        # This way, we can quickly depend on the number of objects returned and compare that
+        # with what the client requested.
+        # For example, if there are 100 objects in the database and the client used limit=100,
+        # it would, internally, ask for 101 objects. So if you retrieved 100 objects
+        # it means we got less than we asked for and thus there is not another page.
+        # Equally, if there are 200 objects in the database and the client used
+        # limit=100 it would, internally, ask for 101 objects and actually get that. Then,
+        # you know there is another page.
+
+        # get_objects() returns an integer if you pass `count_only=True`
+        # and a list if you pass `count_only=False`
+        objects = self.model.get_objects(
             filters=filters,
             sorting=sorting,
-            limit=limit,
+            limit=limit + 1,  # See bigger explanation above.
             pagination_rules=pagination_rules,
             include_deleted=include_deleted,
+            count_only=head_request,
         )
```

Comment:
I'm not a big fan of this pattern. I don't think the return type of a method should differ based on a parameter value. What do you think?